engines/asylum/resources/script.cpp

```diff
@@ -0,0 +1,2102 @@
+	ADD_OPCODE(ChangePlayer);
+	ADD_OPCODE(ChangeActorStatus);
+	ADD_OPCODE(StopSound);
+	ADD_OPCODE(JumpRandom);
+	ADD_OPCODE(ClearScreen);
+	ADD_OPCODE(Quit);
+	ADD_OPCODE(JumpObjectFrame);
+	ADD_OPCODE(DeleteGraphics);
+	ADD_OPCODE(SetActorField944);
+	ADD_OPCODE(SetScriptField1BB0);
+	ADD_OPCODE(OnScriptField1BB0);
+	ADD_OPCODE(Interact);
+	ADD_OPCODE(SetResourcePalette);
+	ADD_OPCODE(SetObjectFrameIndexAndFlags);
+	ADD_OPCODE(SetObjectFlags);
+	ADD_OPCODE(SetActorActionIndex2);
+	ADD_OPCODE(UpdateTransparency);
+	ADD_OPCODE(QueueScript);
+	ADD_OPCODE(ProcessActor);
+	ADD_OPCODE(ClearActorFields);
+	ADD_OPCODE(SetObjectLastFrameIndex);
+	ADD_OPCODE(SetActionAreaFlags);
+	ADD_OPCODE(UpdatePlayerChapter9);
+	ADD_OPCODE(ShowMenu);
+	ADD_OPCODE(UpdateGlobalFlags);
+
+	reset();
+}
+
+ScriptManager::~ScriptManager() {
+	for (int i = 0; i < (int)_opcodes.size(); i++)
+		delete _opcodes[i];
+
+	_scripts.clear();
+	_queue.reset();
+
+	_currentScript = NULL;
+	_currentQueueEntry = NULL;
+
+	// Zero-out passed pointers
+	_vm = NULL;
+}
+
+void ScriptManager::load(Common::SeekableReadStream *stream) {
+	stream->readSint32LE();  // size
+	int32 numEntries = stream->readSint32LE();
+
+	for (int32 a = 0; a < numEntries; a++) {
+		Script script;
+
+		for (int32 c = 0; c < MAX_ACTION_COMMANDS; c++) {
+			ScriptEntry command;
+
+			command.numLines = stream->readSint32LE();
+			command.opcode   = (OpcodeType)stream->readSint32LE();
+			command.param1   = stream->readSint32LE();
+			command.param2   = stream->readSint32LE();
+			command.param3   = stream->readSint32LE();
+			command.param4   = stream->readSint32LE();
+			command.param5   = stream->readSint32LE();
+			command.param6   = stream->readSint32LE();
+			command.param7   = stream->readSint32LE();
+			command.param8   = stream->readSint32LE();
+			command.param9   = stream->readSint32LE();
+
+			script.commands[c] = command;
+
+#ifdef DEBUG
+			// Output encounter info
+			if (command.opcode == kOpcodeRunEncounter)
+				debugC(kDebugLevelEncounter, "%d, %d, %d, %d", command.param1, command.param2, command.param3, command.param4);
+#endif
+		}
+
+		script.field_1BAC = stream->readSint32LE();
+		script.field_1BB0 = stream->readSint32LE();
+		script.counter    = stream->readSint32LE();
+
+		_scripts.push_back(script);
+	}
+}
+
+void ScriptManager::saveLoadWithSerializer(Common::Serializer &s) {
+	for (uint i = 0; i < _scripts.size(); i++)
+		_scripts[i].saveLoadWithSerializer(s);
+}
+
+// Save the script queue (in the original, it is part of the shared data)
+void ScriptManager::saveQueue(Common::Serializer &s) {
+	_queue.saveLoadWithSerializer(s);
+}
+
+void ScriptManager::resetAll() {
+	_scripts.clear();
+	resetQueue();
+	reset();
+}
+
+void ScriptManager::reset(uint32 count) {
+	// Create a set of empty scripts
+	_scripts.clear();
+	for (uint32 i = 0; i < count; i++) {
+		Script script;
+		_scripts.push_back(script);
+	}
+
+	_done = false;
+	_exit = false;
+	_processNextEntry = false;
+
```

Comment:
The following should be added here:
````
    _lastProcessedCmd = NULL;
````

Without this, valgrind indicates another conditional based on uninitialized value at line 406 of this file.